v2/internal/reconcilers/arm/arm_client_cache.go

```diff
@@ -242,3 +257,38 @@ func (c *ARMClientCache) getSecret(ctx context.Context, namespace string, secret
 func matchSecretData(old, new map[string][]byte) bool {
 	return reflect.DeepEqual(old, new)
 }
+
+// TODO: Not sure if its a right thing to put this here or create a separate package for identity/auth and put workloadIdentityCredential there?
+// The below bit is from https://github.com/Azure/azure-sdk-for-go/issues/15615#issuecomment-1211012677
+type workloadIdentityCredential struct {
+	assertion     string
+	tokenFilePath string
+	cred          *azidentity.ClientAssertionCredential
+	lastRead      time.Time
+}
+
+func newWorkloadIdentityCredential(tenantID, clientID, file string) (*workloadIdentityCredential, error) {
+	w := &workloadIdentityCredential{tokenFilePath: file}
+	cred, err := azidentity.NewClientAssertionCredential(tenantID, clientID, w.getAssertion, nil)
+	if err != nil {
+		return nil, err
+	}
+	w.cred = cred
+	return w, nil
+}
+
+func (w *workloadIdentityCredential) GetToken(ctx context.Context, opts policy.TokenRequestOptions) (azcore.AccessToken, error) {
+	return w.cred.GetToken(ctx, opts)
+}
+
+func (w *workloadIdentityCredential) getAssertion(context.Context) (string, error) {
+	if now := time.Now(); w.lastRead.Add(5 * time.Minute).Before(now) {
```

Comment:
Note that there's a data race here, so you may want to add synchronization, if only to avoid setting off the race detector.